src/controller/proxy/local.go

```diff
@@ -0,0 +1,181 @@
+	CheckDependencies(ctx context.Context, man distribution.Manifest, dig string, mediaType string) []distribution.Descriptor
+	// DeleteManifest cleanup delete tag from local cache
+	DeleteManifest(ctx context.Context, repo, ref string)
+}
+
+// local defines operations related to local repo under proxy mode
+type local struct {
+	adapter *base.Adapter
+}
+
+// CreateLocalInterface create the LocalInterface
+func CreateLocalInterface() LocalInterface {
+	return &local{}
+}
+
+// TODO: replace it with head request to local repo
+func (l *local) BlobExist(ctx context.Context, dig string) (bool, error) {
+	return blob.Ctl.Exist(ctx, dig)
+}
+
+func (l *local) init() error {
+	if l.adapter != nil {
+		return nil
+	}
+	registryURL := config.GetCoreURL()
+	reg := &model.Registry{
+		URL: registryURL,
+		Credential: &model.Credential{
+			Type:         model.CredentialTypeSecret,
+			AccessSecret: config.ProxyServiceSecret,
+		},
+	}
+	adapter, err := base.New(reg)
+	l.adapter = adapter
+	return err
+}
+
+func (l *local) PushBlob(ctx context.Context, p *models.Project, localRepo string, desc distribution.Descriptor, bReader io.ReadCloser) error {
+	log.Debugf("Put blob to local registry, localRepo:%v, digest: %v", localRepo, desc.Digest)
+	if err := l.init(); err != nil {
+		return err
+	}
+	err := l.adapter.PushBlob(localRepo, string(desc.Digest), desc.Size, bReader)
+	return err
+}
+
+func (l *local) PushManifest(ctx context.Context, p *models.Project, repo string, tag string, mfst distribution.Manifest) error {
+	// Make sure there is only one go routing to push current artifact to local repo
+	if len(tag) == 0 {
+		// if tag is empty, set to latest
+		tag = "latest"
+	}
+	artifact := repo + ":" + tag
+	mu.Lock()
+	_, ok := inflight[artifact]
+	if ok {
+		mu.Unlock()
+		// Skip to copy artifact if there is existing job running
+		return nil
+	}
+	inflight[artifact] = 1
+	mu.Unlock()
+	defer releaseLock(artifact)
+
+	if err := l.init(); err != nil {
+		return err
+	}
+	mediaType, payload, err := mfst.Payload()
+	if err != nil {
+		return err
+	}
+	_, err = l.adapter.PushManifest(repo, tag, mediaType, payload)
+	return err
+}
+
+// DeleteManifest cleanup delete tag from local cache
+func (l *local) DeleteManifest(ctx context.Context, repo, ref string) {
+	log.Debug("Remove tag from repo if it is exist")
+	// TODO: remove cached tag if it exist in cache
+}
+
+func releaseLock(artifact string) {
+	mu.Lock()
+	delete(inflight, artifact)
+	mu.Unlock()
+}
+
+// updateManifestList -- Trim the manifest list, make sure all depend manifests are ready
+func (l *local) updateManifestList(ctx context.Context, manifest distribution.Manifest) (distribution.Manifest, error) {
+	switch v := manifest.(type) {
+	case *manifestlist.DeserializedManifestList:
+		existMans := make([]manifestlist.ManifestDescriptor, 0)
+		for _, m := range v.Manifests {
+			exist, err := l.BlobExist(ctx, string(m.Digest))
+			if err != nil {
```

Comment:
why continue here?

Reply:
Could you explain the reason why?

I also noticed that l.ManifestExists will never return non-nil error....

Reply:
If one of the depends manifest exist, then we can push the manifest list.

Reply:
Update and return error